docs/designs/allreduce.md

```diff
@@ -1,14 +1,215 @@
+For example, in [ring-allreduce](http://research.baidu.com/bringing-hpc-techniques-deep-learning/) algorithm, each of *N* workers only needs to communicate with two of its peer workers *2 * (N − 1)* times
+to update all the model parameters completely.
+* Scaling up and down of the number of workers is as easy as reconstructing the underlying allreduce communicator and
+re-assigning the ranks among the workers.
+
+We'd like to support allreduce-based distributed training in ElasticDL so our users may obtain more efficient distributed
+training performance in certain use cases without having to worry about many challenges involved in performance tuning and
+resource allocation under parameter-server-based distributed training.
+
+In the following sections, we will explain the design of allreduce-based distributed training in ElasticDL in detail.
+
+For details on the existing technologies relevant to collective communications, please head over to the last section of
+this design doc.
+
+## Design Components
+
+There are two major design components: the fault-tolerant allreduce implementation and support for allreduce-based training in ElasticDL.
+The next two sections will illustrate the details of these components.
+
+### Fault-tolerant Allreduce Implementation
+
+We are collaborating with [Caicloud](https://github.com/caicloud/) on building an API that provides implementations of
+fault-tolerant allreduce. The initial implementation will contain an experimental Python binding for NCCL that is
+fault-tolerant and Kubernetes-native. This will include but not limited to the following objectives (more details to be disclosed later once
+the implementation has been open-sourced):
+
+* Fault-tolerant: if any of the worker pod fails, the [NCCL communicator](https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/usage/communicators.html)
+can be reconstructed. The allreduce operation continues as long as there's at least one healthy worker pod.
+* Elastic: the number of worker pods can be dynamically added if there are enough computational resources available.
+The NCCL communicator will be reconstructed and the ranks can be re-assigned as the number of worker pods changes.
+
+The interface for allreduce would look like the following:
+
+```python
+num_epochs = 2
+num_batches = 10
+data_loader = DataLoader(num_batches)
+communicator = AllreduceCommunicator()
+
+for _ in range(num_epochs):
+    for features, labels in data_loader:
+        outputs = model(features)
+        loss = loss_fn(outputs, labels)
+        grads = calculate_gradients(loss, model)
+        res, averaged_grads = communicator.average_gradients(grads)
+        if res == SUCCESS:
+            update_model(averaged_grads)
+        elif res == FAILED:
+            continue
+```
+
+We also provide API to broadcast all model parameters from one of the worker pods to all other active worker pods.
```

Comment:
If one worker pod restart, which worker will call the broadcast Api?

Reply:
Master selects one worker to call broadcast API.

Reply:
I added "Note that all NCCL API calls are asynchronous but our implementation of the communication collective primitives is synchronous
so it's easier for other frameworks like ElasticDL to use without worrying about failure handling."

cc @zw0610

Reply:
If there is just one pod demanding weights, it seems broadcast is unnecessary. Peer-to-peer communication can cover such occasion, which can take place when the communicator is aware of the new worker join.

Another strategy is append a compulsory broadcast ops after anytime a new communicator is (re-)created, which might use a bit more network bandwidth.

Since the lib is still under development, we can discuss which is preferred.

Reply:
While all workers are created equal, one of them will be considered as a 'master' if self.rank == 0. This worker is responsible for sending weights to the new comers.